File: frontend/src/api/tools.ts
import { API_BASE_URL } from "../config/api";

export type ToolName = "semgrep" | "gitleaks" | "trivy" | "syft" | "grype" | "lynis" | "openscap";
export type UpdatableToolName = Extract<ToolName, "trivy" | "grype">;

export interface ToolStatus {
  installed: boolean;
  version: string | null;
  error: string | null;
}

export type ToolsStatusResponse = Partial<Record<ToolName, ToolStatus>>;

export interface ToolUpdateResponse {
  command: string[];
  success: boolean;
  stdout: string;
  stderr: string;
  exit_code: number | null;
  timed_out: boolean;
  error_message: string | null;
}

async function readErrorDetail(response: Response, fallback: string): Promise<string> {
  try {
    const body = (await response.json()) as { detail?: string };
    if (body.detail) {
      return body.detail;
    }
  } catch {
    // Keep the generic status message when the API does not return JSON.
  }
  return fallback;
}

export async function getToolsStatus(): Promise<ToolsStatusResponse> {
  const response = await fetch(`${API_BASE_URL}/api/tools/status`);

  if (!response.ok) {
    const detail = await readErrorDetail(
      response,
      `Could not load tool status (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ToolsStatusResponse;
}

export async function updateTrivyDb(): Promise<ToolUpdateResponse> {
  const response = await fetch(`${API_BASE_URL}/api/tools/trivy/update-db`, {
    method: "POST",
  });

  if (!response.ok) {
    const detail = await readErrorDetail(
      response,
      `Could not update Trivy DB (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ToolUpdateResponse;
}

export async function updateGrypeDb(): Promise<ToolUpdateResponse> {
  const response = await fetch(`${API_BASE_URL}/api/tools/grype/update-db`, {
    method: "POST",
  });

  if (!response.ok) {
    const detail = await readErrorDetail(
      response,
      `Could not update Grype DB (${response.status})`,
    );
    throw new Error(detail);
  }

  return (await response.json()) as ToolUpdateResponse;
}

File: backend/app/scanners/tools.py
from __future__ import annotations

from dataclasses import dataclass

from app.scanners.runner import run_command


TOOL_STATUS_TIMEOUT_SECONDS = 5


@dataclass(frozen=True)
class ToolStatus:
    installed: bool
    version: str | None
    error: str | None


TOOL_VERSION_COMMANDS = {
    "semgrep": ["semgrep", "--version"],
    "gitleaks": ["gitleaks", "version"],
    "trivy": ["trivy", "--version"],
}


def _version_text(stdout: str, stderr: str) -> str | None:
    text = (stdout or stderr).strip()
    if not text:
        return None
    return text.splitlines()[0].strip() or None


def check_tool_status(tool_name: str, timeout: int | float = TOOL_STATUS_TIMEOUT_SECONDS) -> ToolStatus:
    command = TOOL_VERSION_COMMANDS[tool_name]
    result = run_command(command, timeout=timeout)

    if result.exit_code == 0:
        return ToolStatus(
            installed=True,
            version=_version_text(result.stdout, result.stderr),
            error=None,
        )

    return ToolStatus(
        installed=False,
        version=None,
        error=result.error_message or result.stderr.strip() or f"{tool_name} version check failed",
    )


def get_tools_status() -> dict[str, ToolStatus]:
    return {tool_name: check_tool_status(tool_name) for tool_name in TOOL_VERSION_COMMANDS}
